File: LeetCode/994. Rotting Oranges.java
class Solution {
    public int orangesRotting(int[][] g) {
        int[] DIRS = {1, 0, -1, 0, 1};
        int rows = g.length;
        int cols = g[0].length;

        int res = 0;
        int fresh = 0;
        LinkedList<Pair<Integer,Integer>> q = new LinkedList<>();
        
        for (int r=0; r<rows; r++) {
            for (int c=0; c<cols; c++) {
                Pair<Integer,Integer> coord = new Pair<>(r, c);
                if (g[r][c] == 2) {
                    q.add(coord);
                } else if (g[r][c] == 1) fresh++;
            }
        }

        while (!q.isEmpty() && fresh > 0) {
            int size = q.size();
            for (int i=0; i<size; i++) {
                Pair<Integer,Integer> coord = q.poll();
                int r = coord.getKey();
                int c = coord.getValue();

                for (int d=0; d<DIRS.length-1; d++) {
                    int nr = r + DIRS[d];
                    int nc = c + DIRS[d + 1];
                    Pair<Integer,Integer> newCoord = new Pair<>(nr, nc);
                    if (nr >= 0 && nr < rows &&
                        nc >= 0 && nc < cols &&
                        g[nr][nc] == 1) {
                        q.add(newCoord);
                        g[nr][nc] = 2;
                        fresh--;
                    }
                }
            }
            res++;
        }

        if (fresh > 0) return -1;
        return res;
    }
}

File: LeetCode/994. Rotting Oranges.kt
class Solution {
    var rows = 0
    var cols = 0

    fun orangesRotting(g: Array<IntArray>): Int {
        val q = LinkedList<IntArray>() //queue for the rotten
        var time = 0 //elapsed time
        var cnt = 0 //fresh orange count
        rows = g.size
        cols = g[0].size

        val dirs = arrayOf(intArrayOf(1,0),
                           intArrayOf(-1,0),
                           intArrayOf(0,1),
                           intArrayOf(0,-1)
                           ) //all 4 directions

        for (r in 0 until rows) {
            for (c in 0 until cols) { //loop through all cells
                if (g[r][c] == 1) cnt++ //add to fresh count if it is a fresh orange
                if (g[r][c] == 2) q.add(intArrayOf(r, c)) //add to rotten queue if it is a rotten one
            }
        }

        while (q.isNotEmpty() && cnt > 0) { //while the rotten queue is not empty and fresh oranges still persist
            for (i in 0 until q.size) { //loop current queued rotten oranges
                val rotten = q.poll()
                for(dir in dirs) { //search 4 directions
                    val row = rotten[0] + dir[0]
                    val col = rotten[1] + dir[1]

                    if (row < 0 || row >= rows ||
                        col < 0 || col >= cols ||
                        g[row][col] != 1) { continue } //if the cell is out of bounds or is not fresh, skip

                    g[row][col] = 2
                    q.add(intArrayOf(row, col))
                    cnt-- //make the fresh orange rot, then add to the queue
                }
            }
            time++ //add time for 1 iteration
        }
        return if (cnt == 0) time else -1 //if there still is/are fresh orange(s), return -1, otherwise return elapsed time
    }
}
